### PR TITLE
improve graceful shutdown behaviour #2090

### PR DESCRIPTION
**Pull request by fr3akX**
Fixes behaviour as described in ~~https://github.com/akka/akka/issues/25262~~ https://github.com/akka/akka-http/issues/2090

### CODE REVIEW

**Comment by ktoso** [comment]
Moved issue to the akka-http repo; This refers to https://github.com/akka/akka-http/issues/2090 now

**Comment by fr3akX** [comment]
I added test asserting "Connection: close" and test that validates that connections has been closed.

**Comment by jlprat** [comment]
It's failing due to code formatting differences. Just compile your code locally using `sbt` and you'll see some formatting changes. Commit and push those :)

**Comment by jlprat** [comment]
This is the offending file: `akka-http-core/src/test/scala/akka/http/scaladsl/GracefulTerminationSpec.scala`

**Comment by ktoso** [comment]
Once you click the details for the failed build you'll get to
improve_graceful_shutdown_behaviour__2090_by_fr3akx_ _pull_request__2089_ _akka_akka-http

https://jenkins.akka.io:8498/job/pr-validator-akka-http/2705/console (console output)

```
+ echo '[error] Unformatted code found. Please run '\''test:compile'\'' and commit the reformatted code.'
[error] Unformatted code found. Please run 'test:compile' and commit the reformatted code.
```

Which includes explanation what to do about the error. Please do not commit code that was not compiled :) Compiling takes care of all the proper formatting.

**Comment by ktoso** [comment]
FYI, still reviewing, there's other core concerns before this can be merged.

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala [hunk 1]
```diff
@@ -270,7 +273,11 @@ private[http] final class GracefulTerminatorStage(settings: ServerSettings)
               } else grab(fromUser)
 
             pendingUserHandlerResponse = false
-            push(toNet, response)
+
+            //send inflight response with Connection: close
```
[the remark this answers is not in the document]
Comment (fr3akX) [reply akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala]: can You, please, clarify under which conditions race can occur?

**Comment by lightbend-cla-validator** [comment]
Hi @fr3akX,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by johanandren** [comment]
Not sure what the CLA validator is smoking, fr3akX is in the CLA-signed-db.

**Comment by fr3akX** [comment]
I signed CLA after last validation failed. What else should I do to get this merged?

File: docs/src/main/paradox/server-side/graceful-termination.md [hunk 2]
```diff
@@ -31,15 +31,16 @@ This can be used to signal parts of the application that the HTTP server is shut
 Note also that for more advanced shut down scenarios you may want to use the @extref[Coordinated Shutdown](akka-docs:/actors.html#coordinated-shutdown) capabilities of Akka.
 
 Next, all in flight requests will be handled. If a request is "in-flight" (being handled by user code), it is given `hardDeadline` time to complete.
- 
-- if user code emits a response within the timeout, then this response is sent to the client
+
+- if connection has no "in-flight" request, connection is terminated immediately
```
Comment (ktoso) [inline docs/src/main/paradox/server-side/graceful-termination.md]: Grammar: if a connection has no "in-flight" request, it is terminated immediately

File: docs/src/main/paradox/server-side/graceful-termination.md [hunk 3]
```diff
@@ -31,15 +31,16 @@ This can be used to signal parts of the application that the HTTP server is shut
 Note also that for more advanced shut down scenarios you may want to use the @extref[Coordinated Shutdown](akka-docs:/actors.html#coordinated-shutdown) capabilities of Akka.
 
 Next, all in flight requests will be handled. If a request is "in-flight" (being handled by user code), it is given `hardDeadline` time to complete.
- 
-- if user code emits a response within the timeout, then this response is sent to the client
+
+- if connection has no "in-flight" request, connection is terminated immediately  
+- if user code emits a response within the timeout, then this response is sent to the client with `Connection: close` header and connection is closed.
```
Comment (ktoso) [inline docs/src/main/paradox/server-side/graceful-termination.md]: grammar: with a `Connection: close` header

File: docs/src/main/paradox/server-side/graceful-termination.md [hunk 4]
```diff
@@ -31,15 +31,16 @@ This can be used to signal parts of the application that the HTTP server is shut
...
 - if it is a streaming response, it is also mandated that it shall complete within the deadline, and if it does not
   the connection will be terminated regardless of status of the streaming response. This is because such response could be infinite,
   which could trap the server in a situation where it could not terminate if it were to wait for a response to "finish".
     - existing streaming responses must complete before the deadline as well.
       When the deadline is reached the connection will be terminated regardless of status of the streaming responses.
 - if user code does not reply with a response within the deadline we produce a special @java[`akka.http.javadsl.settings.ServerSettings.getTerminationDeadlineExceededResponse`]@scala[`akka.http.scaladsl.settings.ServerSettings.terminationDeadlineExceededResponse`] 
-HTTP response (e.g. `503 Service Unavailable`)
+HTTP response (e.g. `503 Service Unavailable`) with `Connection: close` and terminate connection.
```
Comment (ktoso) [inline docs/src/main/paradox/server-side/graceful-termination.md]: with a `Connection: close` header an close the connection.

**Comment by fr3akX** [comment]
what else can I do to get this trough? :)

**Comment by johanandren** [comment]
Just waiting for a LGTM from Ktoso now. Requirement is all checks green and two core team LGTMs before merge.

File: akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala [hunk 5]
```diff
@@ -47,9 +47,11 @@ class ServerBinding private[http] (delegate: akka.http.scaladsl.Http.ServerBindi
    * This can be used to signal parts of the application that the http server is shutting down and they should clean up as well.
    * Note also that for more advanced shut down scenarios you may want to use the Coordinated Shutdown capabilities of Akka.
    *
-   * 2) Handle in-flight request:
+   * 2) if a connection has no "in-flight" request, it is terminated immediately
+   *
+   * 3) Handle in-flight request:
    * - if a request is "in-flight" (being handled by user code), it is given `hardDeadline` time to complete,
-   *   - if user code emits a response within the timeout, then this response is sent to the client
+   *   - if user code emits a response within the timeout, then this response is sent to the client with a `Connection: close` header and connection is closed.
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala]: Missing “the”, ‘and the connection is closed’

File: akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala [hunk 6]
```diff
@@ -58,12 +60,12 @@ class ServerBinding private[http] (delegate: akka.http.scaladsl.Http.ServerBindi
    *   - if user code does not reply with a response within the deadline we produce a special [[akka.http.javadsl.settings.ServerSettings.getTerminationDeadlineExceededResponse]]
    *     HTTP response (e.g. 503 Service Unavailable)
    *
-   * 3) Keep draining incoming requests on existing connection:
+   * 4) Keep draining incoming requests on existing connection:
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala]: Yep thanks

**Comment by fr3akX** [comment]
made final grammar adjustments

**Comment by fr3akX** [comment]
Thanks a lot!

**Comment by raboof** [comment]
Thanks for the contribution, great one!

**Comment by ktoso** [comment]
Thanks a lot, sorry for being so picky about wording :)

**Pull request merged**